File: python/Polybius.py
# -*- coding: utf-8 -*-
#
# Polybius Cipher
# 
# @author  lellansin <lellansin@gmail.com>
# @website http://www.lellansin.com/tutorials/ciphers
#
import random

def rand(min, max):
    return int((max - min) * random.random() + min)

# 
# 随机生成棋盘
# 
def generate_table():
    alphabet = 'ABCDEFGHIKLMNOPQRSTUVWXYZ'
    table = [[0] * 5 for row in range(5)]

    for y in range(5):
        for x in range(5):
            table[x][y] = alphabet[rand(0, len(alphabet))]
            alphabet = alphabet.replace(table[x][y], '')
    return table

def getStr(x, format='%02s'):
    return ''.join(format % i for i in x)


# 
# 显示棋盘
# 
def print_table(table):
    print(' ' + getStr(range(1, 6)))
    for row in range(0, len(table)):
        print(str(row + 1) + getStr(table[row]))


# 
# 加密
# 
def encrypt(table, words):
    string = table
    cipher = ''

    for ch in words.upper():
        for row in range(len(table)):
            if ch in table[row]:
                x = str((table[row].index(ch) + 1))
                y = str(row + 1)
                cipher += y + x
    return cipher


# 
# 解密
# 
def decrypt(table, numbers):
    text = ''
    for index in range(0, len(numbers), 2):
        y = int(numbers[index]) - 1
        x = int(numbers[index + 1]) - 1
        text += table[y][x]
    return text


if __name__ == '__main__':
    # 随机生成棋盘
    table = generate_table()
    # 输出棋盘
    print_table(table)
    
    # 使用棋盘加密字符串
    ciphertext = encrypt(table, "hello, world")
    # 输出密文
    print(ciphertext)
    # 解密字符串
    print(decrypt(table, ciphertext))

File: javascript/Polybius.js
/*
 * Polybius Cipher
 *
 * @author  lellansin <lellansin@gmail.com>
 * @website http://www.lellansin.com/tutorials/ciphers
 */
(function() {

    var Polybius = {};

    /*
     * 生成棋盘
     */
    Polybius.generateTable = function() {
        var alphabet = 'ABCDEFGHIKLMNOPQRSTUVWXYZ';
        var table = '';
        var rand;

        while (alphabet.length) {
            rand = alphabet[random(0, alphabet.length)];
            alphabet = alphabet.replace(rand, '');
            table += rand;
        }

        return table;
    };

    /*
     * 打印棋盘
     */
    Polybius.printTable = function(table) {
        var str = '  1 2 3 4 5 \n';
        for (var y = 0; y < 5; y++) {
            str += (y + 1) + ' ';
            for (var x = 0; x < 5; x++) {
                str = str + table[y * 5 + x] + ' ';
            }
            str += '\n';
        }
        console.log(str);
    };

    /*
     * 加密
     */
    Polybius.encrypt = function(table, text) {
        var cipher = '';
        text = text.toUpperCase().replace(/J/g, 'I');

        for (var i = 0; i < text.length; i++) {
            var index = table.indexOf(text[i]);
            if (index >= 0) {
                cipher += (Math.floor(index / 5) + 1) * 10 + (index % 5 + 1);
            }
        }
        return cipher;
    };

    /*
     * 解密
     */
    Polybius.decrypt = function(table, numbers) {
        var result = '';
        var array = numbers.replace(/\s/, '').match(/\d\d/g);

        for (var i = 0; i < array.length; i++) {
            var pos = array[i];
            var y = pos[0] - 1;
            var x = pos[1] - 1;
            result += table[y * 5 + x];
        }

        return result;
    };

    var random = function(min, max) {
        return min + Math.floor(Math.random() * (max - min));
    };

    /* -------------------- 测试 --------------------- */

    var text = 'hello world, this is a encrypt/decrypt test!';

    // 生成棋盘
    var table = Polybius.generateTable();
    Polybius.printTable(table);

    // 加密
    var ciphertext = Polybius.encrypt(table, text);
    console.log(ciphertext);

    // 解密
    console.log(Polybius.decrypt(table, ciphertext));

}());
